### DISCUSSION TITLE
How to add the ssl certificates on both server side and client side of the Piko

### DISCUSSION THREAD

**Discussion by jeevan2nani** [andydunstall/piko#51] [Q&A] [answered]
Hi, I setup the piko server on my server, and using nginx for the piko server since the server ip is not public.

Now on another machine I set up the piko and trying to use piko agent, By providing the server.url as the url which i set up in nginx.

I'm getting the error 
"error":"tls: failed to verify certificate: x509: cannot validate certificate because it doesn't contain any IP SANs"

So where should we configure these ssl certificates?

**Comment by andydunstall** [comment]
Hi @jeevan2nani, the agent does not yet support configuring root CAs, I'll add support for this later today or tomorrow.

So just to confirm, will add agent configuration:
```
tls:
  root_cas: <path>
```
Which overrides the host root CAs to use a custom CA PEM file (flag `--tls.root-cas`).

Does that sound right?

**Reply by jeevan2nani** [reply]
Do we need to use the same CAs on server side as well?

**Reply by andydunstall** [reply, marked as answer]
The Piko server doesn't yet support TLS either (since its usually hosted behind a TLS terminating load balancer) so as long as you've configured Nginx with the a certificate the root CA can verify it should be ok?

I can also add Piko server TLS support if thats something you need

**Reply by jeevan2nani** [reply]
@andydunstall  Thanks for the clarification. If root CA can verify it, it will be ok.

**Reply by andydunstall** [reply]
Hi @jeevan2nani, I've merged https://github.com/andydunstall/piko/pull/53 which adds TLS support for both the agent to configure root CAs, and the server to configure TLS on each port. Let me know if this works for you or if theres anything else you need added

**Reply by jeevan2nani** [reply]
HI @andydunstall  May I know how can I use the CA's with command.
Is it with piko agent --tls.root-cas path?
When I used this it is throwing error, unknown flag

**Reply by andydunstall** [reply]
@jeevan2nani yep `piko agent --tls.root-cas path` is right. You'll have to re-pull main and build the `piko` binary with `make piko` (I haven't tagged a new release with the TLS changes yet)

**Comment by jeevan2nani** [comment]
@andydunstall  
I'm facing the following issue when I'm using my own ca.pem file, and the server-key.pem, server-cert.pem files in my server's nginx.

"error":"tls: failed to verify certificate: x509: certificate signed by unknown authority"

Can you help me out if you know why this issue is comming

**Reply by andydunstall** [reply]
Hi @jeevan2nani, if your getting error `certificate signed by unknown authority` it sounds like you haven't configured `--tls.root-cas` on the agent? Have you configured `piko agent --tls.root-cas ca.pem`?

Its also worth checking that your CA is correct using cURL `--cacert`, such as if you run `curl http://localhost:8001 --cacert ca.pem` do you get the same error?

**Reply by andydunstall** [reply]
Could you also share the full command your using to run the Piko server and agent (including all flags)

**Reply by andydunstall** [reply]
For example testing locally on my machine with TLS enabled on the server upstream port (`piko server --upstream.tls.enabled --upstream.tls.cert ./tmp/localhost.crt --upstream.tls.key ./tmp/localhost.key`)

Using `piko agent my-endpoint/localhost:3000 --server.url https://localhost:8001` fails with `failed to verify certificate: x509: certificate signed by unknown authority`.

But using `piko agent my-endpoint/localhost:3000 --server.url https://localhost:8001 --tls.root-cas ./tmp/myCA.pem` works ok

**Reply by jeevan2nani** [reply]
piko agent my-endpoint/localhost:3001 --tls.root-cas /etc/ssl/cert.pem --server.url https://dev.automator.jio/proxy

**Reply by andydunstall** [reply]
Ok sounds like `--tls.root-cas` needs to point to `ca.pem` if thats the CA you used to sign your server certificate? (Instead of `/etc/ssl/cert.pem`)

**Reply by jeevan2nani** [reply]
@andydunstall  In sever, I'm not using any certs. In server machine I'm just runing the piko server, and in nginx configs I added the server-cert.pem ,server-key.pem files.

In piko agent I'm using the cert.pem which is mounted to docker in the path mentioned in above command

**Reply by andydunstall** [reply]
Ok so just to clarify

Piko agent is connecting to Nginx via TLS which uses certificate/key `server-cert.pem` and `server-key.pem`. And you've configured the agent with `--tls.root-cas` with path to the CA that signed your `server-cert.pem` certificate?

Did you try using `curl https://dev.automator.jio/proxy --cacert <path to CA file>`? Just to check the CA file is correct?

**Reply by jeevan2nani** [reply]
It is giving response endpoint not found

**Reply by andydunstall** [reply]
Can you copy-and-paste the output - is that from cURL or the agent?

**Reply by jeevan2nani** [reply]
/ # curl https://dev.automator.jio/proxy/ --cacert /etc/ssl/cert.pem
{"error":"endpoint not found"}/ #

**Reply by andydunstall** [reply]
Ok cool so it works ok with `curl` - then I'm confused why the same isn't working with agent

Can you just copy and paste the all output your seeing from piko agent

Also it looks like `https://dev.automator.jio/proxy/` is listening on the proxy port rather than the upstream port? (see https://github.com/andydunstall/piko/blob/main/docs/architecture/overview.md#ports for a description of the different server ports)

**Reply by jeevan2nani** [reply]
I tried with upstream port as well. The response is same. Here I just changed the ca file path

{"level":"info","ts":"2024-06-05T10:43:46.797Z","subsystem":"main","msg":"starting piko agent","conf":{"endpoints":[{"id":"my-endpoint","addr":"localhost:3001"}],"server":{"url":"https://dev.automator.jio/proxy/","heartbeat_interval":10000000000,"heartbeat_timeout":10000000000,"reconnect_min_backoff":500000000,"reconnect_max_backoff":15000000000},"auth":{"api_key":""},"forwarder":{"timeout":10000000000},"tls":{"root_cas":"ca.pem"},"admin":{"bind_addr":":9000"},"log":{"level":"info","subsystems":null}}}
{"level":"info","ts":"2024-06-05T10:43:46.8Z","subsystem":"admin.server","msg":"starting http server","addr":"[::]:9000"}
{"level":"info","ts":"2024-06-05T10:43:46.8Z","subsystem":"endpoint","msg":"registering endpoint","endpoint-id":"my-endpoint","forward-addr":"localhost:3001"}
{"level":"warn","ts":"2024-06-05T10:43:46.94Z","subsystem":"endpoint","msg":"failed to connect to server; retrying","endpoint-id":"my-endpoint","url":"wss://dev.automator.jio/piko/v1/listener/my-endpoint","error":"tls: failed to verify certificate: x509: certificate signed by unknown authority"}

**Reply by andydunstall** [reply]
Ok I haven't been able to reproduce the issue

If these are just self-signed certificates for testing, would you be able to upload the nginx config, certificate, key and CA file so I can try to reproduce?

(All `root-cas` is doing is loading the CA file and setting `tls.Config.RootCAs` in Go, so I don't know why that wouldn't work but curl would - are you running both curl and the piko agent in the same environment just to check the CA file is the same in both cases)

**Reply by jeevan2nani** [reply]
Can you clarify me one thing. 
We don't need to set up tls config in piko server right?
If we just provide the tls files in nginx then just piko agent can make handshake and establish the connection right?

One more thing should we expose all 4 ports through nginx config?

**Reply by andydunstall** [reply]
> If we just provide the tls files in nginx then just piko agent can make handshake and establish the connection right?

Thats correct

> One more thing should we expose all 4 ports through nginx config?

It depends what your looking to expose. If the Piko nodes can talk directly then you don't need to expose the gossip port. Its up to you whether you need the other ports to go through Nginx, such as the admin port should not be public.

**Reply by jeevan2nani** [reply]
Nginx Proxy:
server{
listen 443 ssl;
     server_name _;
     ssl_certificate /etc/ssl/certs/server-cert.pem;
     ssl_certificate_key /etc/ssl/certs/server-key.pem;
  location /proxy/ {
        proxy_pass http://localhost:8000/;
  }
  }
  
  
  
  ca files can't able to attach

**Reply by andydunstall** [reply]
Ok thanks I'll have a go at reproducing with Nginx later today

It might also be worth seeing if it works connecting to Piko directly and enabling TLS on the Piko server (without Nginx)

**Reply by jeevan2nani** [reply]
@andydunstall  The issue is resolved. The issue is with the LoadBalancer certs. 
Thanks for your time on this issue.

**Reply by andydunstall** [reply]
@jeevan2nani great!

Let me know if theres anything else you want added to Piko :)

**Comment by skaravad** [comment]
Hello @andydunstall ,
Can the agent support mTLS when connecting to the Piko Server ? we would like to setup robust security between agent and the server , though JWT provides some safety net , we would like to restrict clients who have a valid client cert signed by our own Certificate Authority, this protects Piko Server (TLS can be offloaded to Nginx or some L7 proxy which can take care of mTLS) , so if the Client TLS config can be extended to support TLS auth , it would be really helpful.

Also this can be made optional to enable end-users with optional Cert, Key and CA paths.

**Reply by andydunstall** [reply]
Hey @skaravad, no the Piko agent does not support configuring the client certificates. Nor does the server support configuring client CAs to verify the client certificates, but it sounds like you don't need that anyway

Please open a GH issue describing what you need and I'll aim to get to it soon (it should be a fairly quick change)

**Reply by skaravad** [reply]
hi @andydunstall , thank you for your response. Opened: https://github.com/andydunstall/piko/issues/191
